### PR TITLE
feat: Add options to get version

### PR DESCRIPTION
**Pull request by pandeyanshuman**
**What changed?**

Port `ExecuteWithVersion` and `ExecuteWithMinVersion` options from the Go client ([#1427](https://github.com/cadence-workflow/cadence-go-client/pull/1427), [#1428](https://github.com/cadence-workflow/cadence-go-client/pull/1428)) to the Java client.

- Added `GetVersionOptions` immutable value class with `Builder`, factory methods (`executeWithVersion(int)`, `executeWithMinVersion()`), and validation.
- Added `getVersion` overloads accepting `GetVersionOptions` throughout the call chain: `Workflow` → `WorkflowInternal` → `WorkflowInterceptor` / `WorkflowInterceptorBase` → `SyncDecisionContext` → `DecisionContext` / `DecisionContextImpl` → `ClockDecisionContext`.
- Added convenience methods `Workflow.getVersionWithCustomVersion()` and `Workflow.getVersionWithMinVersion()`.
- Fixed `ClockDecisionContext` to write the actual recorded version (not always `maxSupported`) to the `CadenceChangeVersion` search attribute.

Version selection priority (options only apply when no cached version exists for changeID):
1. Already cached → return cached (options ignored)
2. `customVersion` set → use that version
3. `useMinVersion=true` → use `minSupported`
4. Default → use `maxSupported` (existing behavior, unchanged)

**Why?**

Enables a safe three-step deployment strategy: (1) deploy code supporting both old and new versions, (2) selectively activate the new version via options without a code deploy, (3) retire old code. The Go client already supports this; the Java client needs parity.

**How did you test it?**

- `GetVersionOptionsTest` — 7 unit tests covering builder, factory methods, mutual exclusion validation, DEFAULT_VERSION rejection, and edge cases.
- Compilation verified: `./gradlew compileJava compileTestJava` — BUILD SUCCESSFUL.
- Existing `testGetVersion`, `testGetVersion2`, `testGetVersionWithoutDecisionEvent` should remain green (no behavior change when options are null).

**Potential risks**

- Low risk. All new method overloads use `default` methods on interfaces (`WorkflowInterceptor`, `DecisionContext`) for backward compatibility — existing implementations compile without changes.
- When `options` is `null` (the default path), behavior is identical to the original code.
- The search attribute fix (using recorded version instead of `maxSupported`) changes what gets written to `CadenceChangeVersion` only when options cause a non-maxSupported version to be recorded. Without options, the behavior is unchanged.

**Release notes**

- Added `GetVersionOptions` to control version selection in `Workflow.getVersion()`. New options `executeWithVersion(int)` and `executeWithMinVersion()` allow operators to control which version is recorded on first-write without deploying new code.

**Documentation Changes**

- Updated `CHANGELOG.md` with new entry under `Unreleased`.
- The [Java client versioning docs](https://cadenceworkflow.io/docs/java-client/versioning) is also updated to document the new `GetVersionOptions` parameter, convenience methods, and the three-step deployment strategy through [cadence-workflow/Cadence-Docs#323](https://github.com/cadence-workflow/Cadence-Docs/pull/323)

### CODE REVIEW

File: src/main/java/com/uber/cadence/workflow/GetVersionOptions.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    private Integer customVersion;
+    private boolean useMinVersion;
+
+    /**
+     * Sets a specific version to use when recording a new version marker.
+     *
+     * @param version the version to use; must not be DEFAULT_VERSION (-1)
+     * @throws IllegalArgumentException if version is DEFAULT_VERSION (-1)
+     */
+    public Builder setCustomVersion(int version) {
```
[the remark this answers is not in the document]
Comment (pandeyanshuman) [reply src/main/java/com/uber/cadence/workflow/GetVersionOptions.java:84]: Changed from rejecting only -1 to rejecting all negatives, added `testCustomVersionCannotBeNegative` test

File: src/test/java/com/uber/cadence/workflow/GetVersionOptionsTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/**
```
[the remark this answers is not in the document]
Comment (pandeyanshuman) [reply src/test/java/com/uber/cadence/workflow/GetVersionOptionsTest.java:1]: Added `ClockDecisionContextVersionOptionsTest.java` with 10 tests

File: src/main/java/com/uber/cadence/workflow/WorkflowInterceptor.java [hunk 3]
```diff
@@ -127,6 +127,11 @@ <R> R mutableSideEffect(
 
   int getVersion(String changeID, int minSupported, int maxSupported);
 
+  default int getVersion(
```
[the remark this answers is not in the document]
Comment (pandeyanshuman) [reply src/main/java/com/uber/cadence/workflow/WorkflowInterceptor.java:135]: This is for backward compatibility. Added `@implNote` on default methods in `WorkflowInterceptor.java` and `DecisionContext.java`

File: src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java [hunk 4]
```diff
@@ -320,6 +321,24 @@ private void handleVersionMarker(MarkerRecordedEventAttributes attributes) {
...
+  }
+
+  GetVersionResult getVersion(
+      String changeId,
+      DataConverter converter,
+      int minSupported,
+      int maxSupported,
+      GetVersionOptions options) {
+
+    if (options != null && options.getCustomVersion() != null) {
```
[the remark this answers is not in the document]
Comment (pandeyanshuman) [reply src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java]: Moved `customVersion` range validation inside handler lambda

File: src/main/java/com/uber/cadence/workflow/WorkflowInterceptor.java [hunk 5]
```diff
@@ -127,6 +127,16 @@ <R> R mutableSideEffect(
 
   int getVersion(String changeID, int minSupported, int maxSupported);
 
+  /**
+   * @implNote Custom implementations should override this method to pass {@code options} through
+   *     the interceptor chain. The default delegates to the 3-arg overload, silently dropping
+   *     options.
+   */
+  default int getVersion(
```
[the remark this answers is not in the document]
Comment (pandeyanshuman) [reply src/main/java/com/uber/cadence/workflow/WorkflowInterceptor.java:135]: 1. The default method *does* work correctly; it just ignores the options and uses the existing `maxSupported` behavior, which is the same as passing `null` options
2. Third-party interceptors are rare in this codebase, and the ones that exist (tracing, signal) already have overrides
3. A warning log on every `getVersion` call would be noisy and alarming without being actionable

File: src/main/java/com/uber/cadence/internal/replay/DecisionContext.java [hunk 6]
```diff
@@ -194,6 +195,19 @@ Optional<byte[]> mutableSideEffect(
    */
   int getVersion(String changeID, DataConverter dataConverter, int minSupported, int maxSupported);
 
+  /**
+   * @implNote Custom implementations should override this method to pass {@code options} through.
+   *     The default delegates to the 4-arg overload, silently dropping options.
+   */
+  default int getVersion(
```
[the remark this answers is not in the document]
Comment (pandeyanshuman) [reply src/main/java/com/uber/cadence/internal/replay/DecisionContext.java:202]: 1. The default method *does* work correctly; it just ignores the options and uses the existing `maxSupported` behavior, which is the same as passing `null` options
2. Third-party interceptors are rare in this codebase, and the ones that exist (tracing, signal) already have overrides
3. A warning log on every `getVersion` call would be noisy and alarming without being actionable

**Review by natemort — APPROVED** [review]
